packages/starlight/integrations/expressive-code.ts

```diff
@@ -0,0 +1,265 @@
+		customizeTheme,
+		useThemedSelectionColors = false,
+		styleOverrides: {
+			textMarkers: textMarkersStyleOverrides,
+			...otherStyleOverrides
+		} = {},
+		useStarlightDarkModeSwitch,
+		useStarlightUiThemeColors = config.theme === undefined,
+		minSyntaxHighlightingColorContrast = 5.5,
+		plugins = [],
+		...rest
+	} = config;
+
+	const configuredThemesCount = (Array.isArray(theme) && theme.length) || 1;
+	if (useStarlightDarkModeSwitch !== false) {
+		const loadedThemes: ExpressiveCodeTheme[] = [];
+		const darkModeSwitchCss: string[] = [];
+		plugins.push({
+			name: 'Starlight Dark Mode Switch',
+			baseStyles: ({ theme, themeClassName }) => {
+				loadedThemes.push(theme);
+				darkModeSwitchCss.push(`
+					:root:not([data-theme='${theme.type}']) .${themeClassName} {
+						display: none;
+					}
+				`);
+				if (loadedThemes.length !== configuredThemesCount) return '';
+				const foundDarkThemes = loadedThemes.filter((t) => t.type === 'dark').length;
+				const foundLightThemes = loadedThemes.filter((t) => t.type === 'light').length;
+				const configuredThemesSupportDarkModeSwitch =
+					foundDarkThemes === 1 && foundLightThemes === 1;
+				if (!configuredThemesSupportDarkModeSwitch) {
+					if (useStarlightDarkModeSwitch === true)
+						console.warn(
+							`*** Warning: To support the config option "useStarlightDarkModeSwitch: true", please provide 1 dark and 1 light theme (found ${foundDarkThemes} and ${foundLightThemes}).\n`
+						);
+					return '';
+				}
+				return darkModeSwitchCss.join('\n');
+			},
+		});
+	}
+
+	if (useStarlightUiThemeColors === true && configuredThemesCount < 2)
+		console.warn(`*** Warning: Using the config option "useStarlightUiThemeColors: true" with only one theme is not recommended. For better color contrast, please provide 1 dark and 1 light theme.\n`);
+
+	// Add Expressive Code UI translations (if any) for all defined locales
+	if (locales) {
+		for (const locale in locales) {
+			const lang = locales[locale]?.lang;
+			if (!lang) continue;
+
+			addTranslations(lang, pluginFramesTexts, [
+				'expressiveCode.copyButtonCopied',
+				'expressiveCode.copyButtonTooltip',
+				'expressiveCode.terminalWindowFallbackTitle',
+			]);
+		}
+	}
+
+	return [
+		astroExpressiveCode({
+			theme,
+			customizeTheme: (theme) => {
+				if (useStarlightUiThemeColors) {
+					applyStarlightUiThemeColors(theme);
+				}
+				if (customizeTheme) {
+					theme = customizeTheme(theme) ?? theme;
+				}
+				if (minSyntaxHighlightingColorContrast > 0) {
+					ensureMinTokenColorContrast(theme, minSyntaxHighlightingColorContrast);
+				}
+				return theme;
+			},
+			useThemedSelectionColors,
+			styleOverrides: {
+				codeFontFamily: 'var(--__sl-font-mono)',
+				codeFontSize: 'var(--sl-text-code)',
+				codeLineHeight: 'var(--sl-line-height)',
+				uiFontFamily: 'var(--__sl-font)',
+				textMarkers: {
+					defaultChroma: '45',
+					backgroundOpacity: '60%',
+					...textMarkersStyleOverrides,
+				},
```

Comment:
One more comment here to note that we want to (with the future API coming soon™️) make the default highlight better match the user-configured colour scheme. Currently the default highlight blue and Starlight UI blue don’t quite match:

![UI elements on a Starlight page alongside the code highlight showing different shades of blue](https://github.com/withastro/starlight/assets/357379/4bcab362-32d5-4644-b420-f7a478db1031)

And this could be even more noticeable with a custom colour scheme. While the green/red highlights for insertions and deletions have a semantic purpose, I’m proposing using a highlight colour that is simply a lighter version of the background colour so that it remains consistent no matter the user-configured colour scheme. Here’s a small example thrown together (exact brightness TBD — should maintain accessible contrast with the text):

| With default colour scheme | With custom colour scheme |
|---|---|
| ![image](https://github.com/withastro/starlight/assets/357379/758fed62-1784-4430-9157-8ba05f82b9cf) | ![image](https://github.com/withastro/starlight/assets/357379/6b1f367d-32df-44e8-8954-74887c122680) |

Obviously this would just be the default we set, and users who prefer a specific highlight colour could choose to configure one.

The other option would have been to use the user’s theme accent colour for these highlight’s but I think that risks clashing with the insertion and deletion styles if someone has a red or green theme accent, so this feels like a good neutral default where we only rely on a brightness contrast and not also a hue difference.

Reply:
I like the idea of just changing the lightness of the background color. I experimented with this a bit and found that setting a semi-transparent white or black color as the `mark` background and border seems to already do the trick. You can see the result in the deploy preview, e.g. [here](https://deploy-preview-742--astro-starlight.netlify.app/manual-setup/#configure-the-integration). I like it!